lms/djangoapps/program_enrollments/docs/decisions/0001-program-enrollment-data-integrity.rst

```diff
@@ -0,0 +1,48 @@
+ProgramEnrollment Model Data Integrity
+--------------------------------------
+
+Status
+======
+
+Accepted (circa August 2019)
+
+
+Context
+=======
+
+For the sake of fundamental data integrity, we are introducing 2 unique
+constraints on the ``program_enrollments.ProgramEnrollment`` model, and
+changed the ``external_user_key`` field to be ``NOT NULL`` and included
+a default value for it.
+
+Decisions
+=========
+
+The unique constraints are on the following column sets:
+
+* ``('user', 'external_user_key', 'program_uuid', 'curriculum_uuid')``
+* ``('external_user_key', 'program_uuid', 'curriculum_uuid')``
```

Comment:
Doesn't the second constraint obviate the first? I think in general:
```
(table_is_unique_on(X) and is_subset(X, Y)) implies table_is_unique_on(Y)
```
In this case specifically, "an external key can be used once per program-curriculum" is a more relaxed version of "for each user, an external key can be used once per program-curriculum".